### PR TITLE
 Allow users to pass their own UI via PlatformParamters + ICustomWebUI

### PR DESCRIPTION
**Pull request by bgavrilMS**
The 4 commits can be reviewed individually. 

ICustomWebUI extensibility is enabled on NetDesktop and NetCore. Note that ADAL does not explicitly implement NetCore, so the extensibility has been added to NetStandard. 

Users have the opportunity to add their implementation of ICustomWebUi when they construct the PlatformParamters object. This is similar than MSAL, where they inject the webui when they build the Interactive request object. Also semantically PlatformParamters controls all UI related operations.

### CODE REVIEW

File: devApps/XFormsApp.Droid/XFormsApp.Droid.csproj [hunk 1]
```diff
@@ -37,7 +37,7 @@
     <EmbedAssembliesIntoApk>True</EmbedAssembliesIntoApk>
     <BundleAssemblies>False</BundleAssemblies>
     <AndroidCreatePackagePerAbi>False</AndroidCreatePackagePerAbi>
-    <AndroidSupportedAbis>armeabi,armeabi-v7a,x86</AndroidSupportedAbis>
+    <AndroidSupportedAbis>armeabi-v7a,x86</AndroidSupportedAbis>
```
Comment (bgavrilMS) [inline devApps/XFormsApp.Droid/XFormsApp.Droid.csproj]: Unrelated change: armeabi has been deprecated starting in VS2019. No impact on VS2017.

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/UI/CustomWebUiHandler.cs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+using System.Threading.Tasks;
+using Microsoft.Identity.Core.Helpers;
+using Microsoft.Identity.Core.Http;
+using Microsoft.IdentityModel.Clients.ActiveDirectory;
+using Microsoft.IdentityModel.Clients.ActiveDirectory.Extensibility;
+using Microsoft.IdentityModel.Clients.ActiveDirectory.Internal.OAuth2;
+
+namespace Microsoft.Identity.Core.UI
+{
+    internal class CustomWebUiHandler : IWebUI
```
Comment (bgavrilMS) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/UI/CustomWebUiHandler.cs:41]: This is identical to MSAL impl.

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/android/PlatformParameters.cs [hunk 3]
```diff
@@ -82,6 +83,14 @@ public PlatformParameters(Activity callerActivity, bool useBroker, PromptBehavio
             PromptBehavior = promptBehavior;
         }
 
+        /// <summary>
+        /// Constructor that allows extends to configure their own web ui. Not supported on Android, iOS and UWP. 
+        /// </summary>
+        public PlatformParameters(PromptBehavior promptBehavior, ICustomWebUi customWebUi)
```
Comment (bgavrilMS) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/android/PlatformParameters.cs:89]: As discussed with @jmprieur , we will expose this extensibility point via PlatformParamters. I have added this ctor to all platforms and on Android, iOS and UWP I am throwing a PlatformNotSupported. This is to avoid having MissingMethodException, because for ADAL we do not have a pure NetCore impl, so users targetting XForms, for example, will see this ctor.

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/net45/WebUIFactory.cs [hunk 4]
```diff
@@ -34,9 +34,13 @@ internal class WebUIFactory : IWebUIFactory
     {
         public IWebUI CreateAuthenticationDialog(CoreUIParent coreUIParent, RequestContext context)
         {
-            if (coreUIParent.UseHiddenBrowser)
+            if (coreUIParent.CustomWebUi != null )
```
Comment (bgavrilMS) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/net45/WebUIFactory.cs:37]: General rule when configuring PlatformParamters - if you have set a CustomWebUI, use that. Otherwise, use existing logic to determine which WebUI you want to use.

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/netcore/NetCoreLegacyCachePersistance.cs [hunk 5]
```diff
@@ -1,46 +0,0 @@
-﻿//----------------------------------------------------------------------
```
Comment (bgavrilMS) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/netcore/NetCoreLegacyCachePersistance.cs:1]: All these netcore files were leftover from the ADAL+MSAL unified repo. They are not in use.

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/netstandard1.3/PlatformInformation.cs [hunk 6]
```diff
@@ -25,9 +25,40 @@
...
+                case PromptBehavior.Always:
+                    authorizationRequestParameters[OAuthParameter.Prompt] = PromptValue.Login;
+                    break;
+                case PromptBehavior.SelectAccount:
+                    authorizationRequestParameters[OAuthParameter.Prompt] = PromptValue.SelectAccount;
+                    break;
+                case PromptBehavior.RefreshSession:
+                    authorizationRequestParameters[OAuthParameter.Prompt] = PromptValue.RefreshSession;
+                    break;
+                case PromptBehavior.Never:  //TODO: needs testing
```
Comment (bgavrilMS) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/netstandard1.3/PlatformInformation.cs:57]: Still testing this case :)

File: tests/Test.ADAL.Integration/SeleniumTests/InteractiveFlowTests.cs [hunk 7]
```diff
@@ -129,11 +129,16 @@ private async Task RunTestForUserAsync(LabResponse labResponse)
...
 
             // Act
-            AuthenticationResult result = await context.AcquireTokenAsync(AdalTestConstants.MSGraph, labResponse.AppId, new Uri(SeleniumWebUIFactory.FindFreeLocalhostRedirectUri()), new PlatformParameters(PromptBehavior.Always));
+            AuthenticationResult result = await context.AcquireTokenAsync(
+                AdalTestConstants.MSGraph, 
+                labResponse.AppId, 
+                new Uri(SeleniumWebUI.FindFreeLocalhostRedirectUri()),
+                new PlatformParameters(
+                    PromptBehavior.Always, 
+                    new SeleniumWebUI(seleniumLogic, _seleniumTimeout)));
```
Comment (bgavrilMS) [inline tests/Test.ADAL.Integration/SeleniumTests/InteractiveFlowTests.cs:139]: Example of the API is use.

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Extensibility/ICustomWebUI.cs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    /// <summary>
+    ///     Interface that an ADAL.NET extender can implement to provide their own Web UI in public client applications
+    ///     to sign-in user and have them consented part of the Authorization code flow.
+    ///     ADAL.NET provides an embedded web view for Windows, but there are other scenarios not yet supported.
+    ///     This extensibility point enables them to provide such UI in a secure way
+    /// </summary>
+    public interface ICustomWebUi
+    {
+        /// <summary>
+        ///     Method called by ADAL.NET to delegate the authentication code Web with with the STS
```
Comment (jmprieur) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Extensibility/ICustomWebUI.cs]: Is there a reason for the indentation?
Apparently this perturbates the Reference doc generator:
look at the remarks section in https://docs.microsoft.com/en-us/dotnet/api/microsoft.identity.client.extensibility.icustomwebui.acquireauthorizationcodeasync?view=azure-dotnet#Microsoft_Identity_Client_Extensibility_ICustomWebUi_AcquireAuthorizationCodeAsync_System_Uri_System_Uri_

**Review by jmprieur — COMMENTED** [review]
LGTM.
One question about XML coments (I changed them in MSAL)

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Extensibility/ICustomWebUI.cs (hunk 8 above)
Comment (bgavrilMS) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Extensibility/ICustomWebUI.cs]: No, I just copy pasted the comments from MSAL. Will fix here and also on MSAL.

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/Exceptions/CoreErrorMessages.cs [hunk 9]
```diff
@@ -115,5 +117,14 @@ internal class CoreErrorMessages
 
         public const string NoRedirectUri = "No redirectUri was configured. ADAL does not provide any defaults.";
 
+        public const string CustomWebUiReturnedInvalidUri = "ICustomWebUi returned an invalid uri - it is empty or has no query.";
+
+        public static string CustomWebUiRedirectUriMismatch(string expectedUri, string actualUri)
```
Comment (MarkZuber) [inline src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/Exceptions/CoreErrorMessages.cs:122]: Since you're in here, can you fix the other error messages above (like MissingEntitlements and DefaultRedirectUriIsINvalid) to be functions like this one instead of formatting them out of band?  thanks!

Comment (MarkZuber) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Core/Exceptions/CoreErrorMessages.cs:122]: oh, this is ADAL.  you can still do it if you want, but not as important since there's a lot of other legacy stuff related to this.

---
In reply to: [262185082](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1511#discussion_r262185082) [](ancestors = 262185082)

File: src/Microsoft.IdentityModel.Clients.ActiveDirectory/Extensibility/ICustomWebUI.cs (hunk 8 above)
Comment (MarkZuber) [reply src/Microsoft.IdentityModel.Clients.ActiveDirectory/Extensibility/ICustomWebUI.cs]: This was a remnant of resharper autoformatting in the comments.  as we move to vs2019 and editorconfig this won't get repeated any longer.

**Review by MarkZuber — APPROVED** [review]
:shipit:
